### PR TITLE
[SPARK-19459][SQL] Add Hive datatype (char/varchar) to StructField metadata

### PR DESCRIPTION
**Pull request by hvanhovell**
## What changes were proposed in this pull request?
Reading from an existing ORC table which contains `char` or `varchar` columns can fail with a `ClassCastException` if the table metadata has been created using Spark. This is caused by the fact that spark internally replaces `char` and `varchar` columns with a `string` column.

This PR fixes this by adding the hive type to the `StructField's` metadata under the `HIVE_TYPE_STRING` key. This is picked up by the `HiveClient` and the ORC reader, see https://github.com/apache/spark/pull/16060 for more details on how the metadata is used.

## How was this patch tested?
Added a regression test to `OrcSourceSuite`.

### CODE REVIEW

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcSourceSuite.scala [hunk 1]
```diff
@@ -162,6 +162,27 @@ abstract class OrcSuite extends QueryTest with TestHiveSingleton with BeforeAndA
       hiveClient.runSqlHive("DROP TABLE IF EXISTS orc_varchar")
     }
   }
+
+  test("read varchar column from orc tables created by hive") {
+    try {
+      // This is an ORC file with a single VARCHAR(10) column that's created using Hive 1.2.1
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcSourceSuite.scala]: Hi, @hvanhovell .
Nit. It's three columns.
```
Structure for orc/orc_text_types.orc
File Version: 0.12 with HIVE_8732
Rows: 1
Compression: ZLIB
Compression size: 262144
Type: struct<_col0:string,_col1:char(10),_col2:varchar(10)>
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 2]
```diff
@@ -1457,8 +1457,31 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with Logging {
...
+    val builder = new MetadataBuilder
+    // Add comment to metadata
+    if (STRING != null) {
+      builder.putString("comment", string(STRING))
+    }
+    // Add Hive type string to metadata.
+    dataType match {
+      case p: PrimitiveDataTypeContext =>
+        val dt = p.identifier.getText.toLowerCase
+        (dt, p.INTEGER_VALUE().asScala.toList) match {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: nit:
```
p.identifier.getText.toLowerCase match {
  case "varchar" | "char" => builder.putString(HIVE_TYPE_STRING, dataType.getText.toLowerCase)
}
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/package.scala [hunk 3]
```diff
@@ -21,4 +21,10 @@ package org.apache.spark.sql
  * Contains a type system for attributes produced by relations, including complex types like
  * structs, arrays and maps.
  */
-package object types
+package object types {
+  /**
+   * Metadata key used to store the Hive type name. This is relevant for datatypes that do not
+   * have a direct Spark SQL counterpart, such as CHAR and VARCHAR.
+   */
+  val HIVE_TYPE_STRING = "HIVE_TYPE_STRING"
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/package.scala]: shall we remove `HiveUtils. HIVE_TYPE_STRING`?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcSourceSuite.scala [hunk 4]
```diff
@@ -162,6 +162,28 @@ abstract class OrcSuite extends QueryTest with TestHiveSingleton with BeforeAndA
       hiveClient.runSqlHive("DROP TABLE IF EXISTS orc_varchar")
     }
   }
+
+  test("read varchar column from orc tables created by hive") {
+    try {
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcSourceSuite.scala]: how about
```
    val hiveClient = spark.sharedState.externalCatalog.asInstanceOf[HiveExternalCatalog].client
     try {
       hiveClient.runSqlHive("CREATE TABLE hive_orc(a VARCHAR(10)) STORED AS orc LOCATION xxx")
       hiveClient.runSqlHive("INSERT INTO TABLE orc_varchar SELECT 'a' FROM (SELECT 1) t")
       sql("CREATE EXTERNAL TABLE spark_orc ...")
       checkAnswer...
     } finally {
        sql("DROP TABLE IF EXISTS ...")
        ...
      }
```
then we don't need to create the orc file manually.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/package.scala (hunk 3 above)
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/package.scala]: Yeah we should.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/MetastoreRelation.scala [hunk 5]
```diff
@@ -32,7 +32,7 @@ import org.apache.spark.sql.catalyst.catalog._
 import org.apache.spark.sql.catalyst.expressions.{AttributeMap, AttributeReference, Expression}
 import org.apache.spark.sql.catalyst.plans.logical.{LeafNode, LogicalPlan, Statistics}
 import org.apache.spark.sql.execution.FileRelation
-import org.apache.spark.sql.types.StructField
+import org.apache.spark.sql.types._
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/MetastoreRelation.scala]: Unnecessary change?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala [hunk 6]
```diff
@@ -51,6 +51,9 @@ private[hive] case class HiveSimpleUDF(
   @transient
   lazy val function = funcWrapper.createFunction[UDF]()
 
+  {
+    function
+  }
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala]: What is the reason for this?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/package.scala [hunk 7]
```diff
@@ -21,4 +21,12 @@ package org.apache.spark.sql
  * Contains a type system for attributes produced by relations, including complex types like
  * structs, arrays and maps.
  */
-package object types
+package object types {
+  /**
+   * Metadata key used to store the the raw hive type string in the metadata of StructField. This
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/package.scala]: Nit: `the the` -> `the`

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcSourceSuite.scala [hunk 8]
```diff
@@ -162,6 +162,40 @@ abstract class OrcSuite extends QueryTest with TestHiveSingleton with BeforeAndA
...
+           |  a STRING,
+           |  b CHAR(10),
+           |  c VARCHAR(10))
+           |STORED AS orc""".stripMargin)
+      // Hive throws an exception if I assign the location in the create table statment.
+      hiveClient.runSqlHive(
+        s"ALTER TABLE hive_orc SET LOCATION '$location'")
+      hiveClient.runSqlHive(
+        "INSERT INTO TABLE hive_orc SELECT 'a', 'b', 'c' FROM (SELECT 1) t")
+
```
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcSourceSuite.scala]: How about adding one more check?

```Scala
      checkAnswer(spark.table("hive_orc"), Row("a", "b         ", "c"))
```

Then, we can remove the test case `SPARK-18220: read Hive orc table with varchar column`

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala (hunk 6 above)
Comment (hvanhovell) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala]: That is my bad.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/MetastoreRelation.scala (hunk 5 above)
Comment (hvanhovell) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/MetastoreRelation.scala]: That just makes it easier to use `HIVE_TYPE_STRING`.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcSourceSuite.scala (hunk 8 above)
Comment (hvanhovell) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcSourceSuite.scala]: yeah that makes sense

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/package.scala (hunk 7 above)
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/package.scala]: will do

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcSourceSuite.scala (hunk 8 above)
Comment (hvanhovell) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcSourceSuite.scala]: Done.

**Comment by hvanhovell** [comment]
retest this please

**Comment by gatorsmile** [comment]
LGTM pending test

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcSourceSuite.scala [hunk 9]
```diff
@@ -152,14 +152,39 @@ abstract class OrcSuite extends QueryTest with TestHiveSingleton with BeforeAndA
     assert(new OrcOptions(Map("Orc.Compress" -> "NONE")).compressionCodec == "NONE")
   }
 
-  test("SPARK-18220: read Hive orc table with varchar column") {
+  test("SPARK-19459/SPARK-18220: read char/varchar column written by Hive") {
     val hiveClient = spark.sharedState.externalCatalog.asInstanceOf[HiveExternalCatalog].client
+    val location = Utils.createTempDir().toURI
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcSourceSuite.scala]: shall we remove this temp dir in the finally block?

**Comment by cloud-fan** [comment]
LGTM, merging to master!

**Comment by weiatwork** [comment]
This doesn't solve the problem when reading a CHAR/VARCHAR column in Hive from a table created using Spark, does it? Hive will fail when trying to convert the String to its CHAR/VARCHAR type
